cmd/auth/main.go

```diff
@@ -250,40 +259,83 @@ func newService(db *sqlx.DB, tracer opentracing.Tracer, secret string, logger lo
-		errs <- http.ListenAndServeTLS(p, certFile, keyFile, httpapi.MakeHandler(svc, tracer))
-		return
+		go func() {
+			errCh <- server.ListenAndServeTLS(certFile, keyFile)
+		}()
+		protocol = "https"
+	default:
+		logger.Info(fmt.Sprintf("Authentication service started using http, exposed port %s", port))
+		go func() {
+			errCh <- server.ListenAndServe()
+		}()
 	}
-	logger.Info(fmt.Sprintf("Authentication service started using http, exposed port %s", port))
-	errs <- http.ListenAndServe(p, httpapi.MakeHandler(svc, tracer))
 
+	select {
+	case <-ctx.Done():
+		ctxShutDown, cancelShutDown := context.WithTimeout(context.Background(), graceWaitTIme*time.Second)
+		defer cancelShutDown()
+		if err := server.Shutdown(ctxShutDown); err != nil {
+			logger.Error(fmt.Sprintf("Authentication %s service error occurred during shutdown at %s: %s", protocol, p, err))
+			return fmt.Errorf("Authentication %s service error occurred during shutdown at %s: %w", protocol, p, err)
+		}
+		logger.Info(fmt.Sprintf("Authentication %s service shutdown of http at %s", protocol, p))
+		return nil
+	case err := <-errCh:
+		return err
+	}
 }
 
-func startGRPCServer(tracer opentracing.Tracer, svc auth.Service, port string, certFile string, keyFile string, logger logger.Logger, errs chan error) {
+func startGRPCServer(ctx context.Context, tracer opentracing.Tracer, svc auth.Service, port string, certFile string, keyFile string, logger logger.Logger) error {
 	p := fmt.Sprintf(":%s", port)
+	errCh := make(chan error)
+
 	listener, err := net.Listen("tcp", p)
 	if err != nil {
-		logger.Error(fmt.Sprintf("Failed to listen on port %s: %s", port, err))
+		return fmt.Errorf("failed to listen on port %s: %w", port, err)
 	}
 
 	var server *grpc.Server
-	if certFile != "" || keyFile != "" {
+	switch {
+	case certFile != "" || keyFile != "":
 		creds, err := credentials.NewServerTLSFromFile(certFile, keyFile)
 		if err != nil {
-			logger.Error(fmt.Sprintf("Failed to load auth certificates: %s", err))
-			os.Exit(1)
+			return fmt.Errorf("failed to load auth certificates: %w", err)
 		}
 		logger.Info(fmt.Sprintf("Authentication gRPC service started using https on port %s with cert %s key %s", port, certFile, keyFile))
 		server = grpc.NewServer(grpc.Creds(creds))
-	} else {
+	default:
 		logger.Info(fmt.Sprintf("Authentication gRPC service started using http on port %s", port))
 		server = grpc.NewServer()
 	}
 
 	mainflux.RegisterAuthServiceServer(server, grpcapi.NewServer(tracer, svc))
 	logger.Info(fmt.Sprintf("Authentication gRPC service started, exposed port %s", port))
-	errs <- server.Serve(listener)
+	go func() {
+		errCh <- server.Serve(listener)
+	}()
+
+	select {
+	case <-ctx.Done():
+		c := make(chan bool)
+		go func() {
+			defer close(c)
+			server.GracefulStop()
+		}()
+		select {
+		case <-c:
+		case <-time.After(graceWaitTIme * time.Second):
```

Comment:
`graceWaitTime` constant can already be calculated like `5*time.Second` so that you use it like `time.After(graceWaitTime)`. I agree with @drasko, maybe simple `stopTimeout` is a better naming.